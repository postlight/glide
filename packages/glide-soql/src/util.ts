export const IS_EMPTY = Symbol("glide.isEmpty");

export interface Fragment {
  [IS_EMPTY](): boolean;
  toString(): string;
}

export function notEmpty(input: Fragment): boolean {
  return !isEmpty(input);
}

export function isEmpty(input: Fragment): boolean {
  return input[IS_EMPTY]();
}
