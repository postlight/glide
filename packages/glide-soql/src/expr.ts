import * as ops from "./ops";
import { IS_EMPTY, Fragment } from "./util";

const SINGLE_QUOTE = /'/gm;

export type Expr = Bool | Cmp;
export type Field = string | Fn;
export type Scalar = bigint | boolean | null | number | string;

export class Bool implements Fragment {
  private readonly value: Expr[];

  private constructor(private readonly operator: ops.Bool, value: Expr[]) {
    this.value = value.reduce<Expr[]>((acc, item) => {
      if (item instanceof Cmp || operator !== item.operator) {
        acc.push(item);
        return acc;
      }

      return acc.concat(item.value);
    }, []);
  }

  static and(...value: Expr[]): Bool {
    return new Bool(ops.Bool.And, value);
  }

  static or(...value: Expr[]): Bool {
    return new Bool(ops.Bool.Or, value);
  }

  [IS_EMPTY](): boolean {
    return this.value.length === 0;
  }

  toString(): string {
    return this.value
      .map(expr => (expr instanceof Bool ? `(${expr})` : `${expr}`))
      .join(` ${this.operator} `);
  }
}

export class Cmp implements Fragment {
  private constructor(
    private readonly field: Field,
    private readonly operator: ops.Cmp,
    private readonly expression: Scalar | Scalar[],
  ) {}

  static eq(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Eq, expression);
  }

  static gt(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Gt, expression);
  }

  static gte(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Gte, expression);
  }

  static in(field: Field, expression: Scalar[]): Cmp {
    return new Cmp(field, ops.Cmp.In, expression);
  }

  static like(field: Field, expression: string): Cmp {
    return new Cmp(field, ops.Cmp.Like, expression);
  }

  static lt(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Lt, expression);
  }

  static lte(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Lte, expression);
  }

  static ne(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Ne, expression);
  }

  static notIn(field: Field, expression: Scalar[]): Cmp {
    return new Cmp(field, ops.Cmp.NotIn, expression);
  }

  [IS_EMPTY](): boolean {
    return false;
  }

  toString(): string {
    let expression: string;

    if (Array.isArray(this.expression)) {
      expression = `(${this.expression.map(stringifyScalar).join(", ")})`;
    } else {
      expression = stringifyScalar(this.expression);
    }

    return `${this.field} ${this.operator} ${expression}`;
  }
}

export class Fn implements Fragment {
  private constructor(private readonly ident: string, private readonly field?: string) {}

  static avg(field: string): Fn {
    return new Fn("AVG", field);
  }

  static calendarMonth(field: string): Fn {
    return new Fn("CALENDAR_MONTH", field);
  }

  static calendarQuarter(field: string): Fn {
    return new Fn("CALENDAR_QUARTER", field);
  }

  static calendarYear(field: string): Fn {
    return new Fn("CALENDAR_YEAR", field);
  }

  static convertTimezone(field: string): Fn {
    return new Fn("CONVERT_TIMEZONE", field);
  }

  static count(field?: string): Fn {
    return new Fn("COUNT", field);
  }

  static countDistinct(field?: string): Fn {
    return new Fn("COUNT_DISTINCT", field);
  }

  static dayInMonth(field: string): Fn {
    return new Fn("DAY_IN_MONTH", field);
  }

  static dayInWeek(field: string): Fn {
    return new Fn("DAY_IN_WEEK", field);
  }

  static dayInYear(field: string): Fn {
    return new Fn("DAY_IN_YEAR", field);
  }

  static dayOnly(field: string): Fn {
    return new Fn("DAY_ONLY", field);
  }

  static fiscalMonth(field: string): Fn {
    return new Fn("FISCAL_MONTH", field);
  }

  static fiscalQuarter(field: string): Fn {
    return new Fn("FISCAL_QUARTER", field);
  }

  static fiscalYear(field: string): Fn {
    return new Fn("FISCAL_YEAR", field);
  }

  static hourInDay(field: string): Fn {
    return new Fn("HOUR_IN_DAY", field);
  }

  static max(field: string): Fn {
    return new Fn("MAX", field);
  }

  static min(field: string): Fn {
    return new Fn("MIN", field);
  }

  static sum(field: string): Fn {
    return new Fn("SUM", field);
  }

  static weekInMonth(field: string): Fn {
    return new Fn("WEEK_IN_MONTH", field);
  }

  static weekInYear(field: string): Fn {
    return new Fn("WEEK_IN_YEAR", field);
  }

  [IS_EMPTY](): boolean {
    return false;
  }

  toString(): string {
    return `${this.ident}(${this.field ? this.field : ""})`;
  }
}

function stringifyScalar(input: Scalar): string {
  switch (typeof input) {
    case "boolean":
      return `${input}`.toUpperCase();
    case "string":
      return `'${input.replace(SINGLE_QUOTE, "\\'")}'`;
    default:
      return input == null ? "null" : String(input);
  }
}
