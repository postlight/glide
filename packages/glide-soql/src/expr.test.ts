import { Bool, Cmp, Fn } from "./expr";
import { isEmpty } from "./util";

describe("Bool", () => {
  test("constructor()", () => {
    const flat = Bool.and(
      Cmp.eq("EmailAddress", "hello@postlight.com"),
      Bool.and(Cmp.eq("FirstName", "Hello"), Cmp.eq("LastName", "Postlight")),
    );

    const nested = Bool.and(
      Cmp.eq("EmailAddress", "hello@postlight.com"),
      Bool.or(Cmp.eq("FirstName", "Hello"), Cmp.eq("LastName", "Postlight")),
    );

    expect(flat).toMatchSnapshot();
    expect(nested).toMatchSnapshot();
  });

  test("#@@isEmpty()", () => {
    const condition = Cmp.eq("EmailAddress", "hello@postlight.com");

    expect(isEmpty(Bool.and())).toBe(true);
    expect(isEmpty(Bool.and(condition))).toBe(false);
  });

  test("#toString()", () => {
    const subject = Bool.and(
      Cmp.eq("EmailAddress", "hello@postlight.com"),
      Bool.or(Cmp.eq("FirstName", "Hello"), Cmp.eq("LastName", "Postlight")),
    );

    expect(`${subject}`).toMatchSnapshot();
  });
});

describe("Cmp", () => {
  test("#@@isEmpty()", () => {
    expect(isEmpty(Cmp.eq("Id", null))).toBe(false);
  });

  test("#toString()", () => {
    expect(`${Cmp.eq("Id", null)}`).toMatchSnapshot();
    expect(`${Cmp.eq("Active", true)}`).toMatchSnapshot();
    expect(`${Cmp.gt("UsedSpace", BigInt(1e12))}`).toMatchSnapshot();
    expect(`${Cmp.gte("UsedSpace", BigInt(1e12))}`).toMatchSnapshot();
    expect(`${Cmp.in("Company", ["Postlight", "Salesforce"])}`).toMatchSnapshot();
    expect(`${Cmp.like("Company", "%Post%")}`).toMatchSnapshot();
    expect(`${Cmp.ne("Description", "'test'")}`).toMatchSnapshot();
    expect(`${Cmp.notIn("Company", ["Postlight", "Salesforce"])}`).toMatchSnapshot();
    expect(`${Cmp.lt("UsedSpace", 1e12)}`).toMatchSnapshot();
    expect(`${Cmp.lte("UsedSpace", 1e12)}`).toMatchSnapshot();
  });
});

describe("Fn", () => {
  test("#@@isEmpty()", () => {
    expect(isEmpty(Fn.count())).toBe(false);
  });

  test("#toString()", () => {
    expect(`${Fn.avg("UsedSpace")}`).toMatchSnapshot();
    expect(`${Fn.calendarMonth("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.calendarQuarter("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.calendarYear("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.convertTimezone("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.count()}`).toMatchSnapshot();
    expect(`${Fn.count("Id")}`).toMatchSnapshot();
    expect(`${Fn.countDistinct()}`).toMatchSnapshot();
    expect(`${Fn.countDistinct("Id")}`).toMatchSnapshot();
    expect(`${Fn.dayInMonth("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.dayInWeek("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.dayInYear("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.dayOnly("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.fiscalMonth("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.fiscalQuarter("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.fiscalYear("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.hourInDay("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.max("max")}`).toMatchSnapshot();
    expect(`${Fn.min("min")}`).toMatchSnapshot();
    expect(`${Fn.sum("sum")}`).toMatchSnapshot();
    expect(`${Fn.weekInMonth("LastLogin")}`).toMatchSnapshot();
    expect(`${Fn.weekInYear("LastLogin")}`).toMatchSnapshot();
  });
});
