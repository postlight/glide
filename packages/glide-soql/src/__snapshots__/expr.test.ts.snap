// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bool #toString() 1`] = `"EmailAddress = 'hello@postlight.com' AND (FirstName = 'Hello' OR LastName = 'Postlight')"`;

exports[`Bool constructor() 1`] = `
Bool {
  "operator": "AND",
  "value": Array [
    Cmp {
      "expression": "hello@postlight.com",
      "field": "EmailAddress",
      "operator": "=",
    },
    Cmp {
      "expression": "Hello",
      "field": "FirstName",
      "operator": "=",
    },
    Cmp {
      "expression": "Postlight",
      "field": "LastName",
      "operator": "=",
    },
  ],
}
`;

exports[`Bool constructor() 2`] = `
Bool {
  "operator": "AND",
  "value": Array [
    Cmp {
      "expression": "hello@postlight.com",
      "field": "EmailAddress",
      "operator": "=",
    },
    Bool {
      "operator": "OR",
      "value": Array [
        Cmp {
          "expression": "Hello",
          "field": "FirstName",
          "operator": "=",
        },
        Cmp {
          "expression": "Postlight",
          "field": "LastName",
          "operator": "=",
        },
      ],
    },
  ],
}
`;

exports[`Cmp #toString() 1`] = `"Id = null"`;

exports[`Cmp #toString() 2`] = `"Active = TRUE"`;

exports[`Cmp #toString() 3`] = `"UsedSpace > 1000000000000"`;

exports[`Cmp #toString() 4`] = `"UsedSpace >= 1000000000000"`;

exports[`Cmp #toString() 5`] = `"Company IN ('Postlight', 'Salesforce')"`;

exports[`Cmp #toString() 6`] = `"Company LIKE '%Post%'"`;

exports[`Cmp #toString() 7`] = `"Description != '\\\\'test\\\\''"`;

exports[`Cmp #toString() 8`] = `"Company NOT IN ('Postlight', 'Salesforce')"`;

exports[`Cmp #toString() 9`] = `"UsedSpace < 1000000000000"`;

exports[`Cmp #toString() 10`] = `"UsedSpace <= 1000000000000"`;

exports[`Fn #toString() 1`] = `"AVG(UsedSpace)"`;

exports[`Fn #toString() 2`] = `"CALENDAR_MONTH(LastLogin)"`;

exports[`Fn #toString() 3`] = `"CALENDAR_QUARTER(LastLogin)"`;

exports[`Fn #toString() 4`] = `"CALENDAR_YEAR(LastLogin)"`;

exports[`Fn #toString() 5`] = `"CONVERT_TIMEZONE(LastLogin)"`;

exports[`Fn #toString() 6`] = `"COUNT()"`;

exports[`Fn #toString() 7`] = `"COUNT(Id)"`;

exports[`Fn #toString() 8`] = `"COUNT_DISTINCT()"`;

exports[`Fn #toString() 9`] = `"COUNT_DISTINCT(Id)"`;

exports[`Fn #toString() 10`] = `"DAY_IN_MONTH(LastLogin)"`;

exports[`Fn #toString() 11`] = `"DAY_IN_WEEK(LastLogin)"`;

exports[`Fn #toString() 12`] = `"DAY_IN_YEAR(LastLogin)"`;

exports[`Fn #toString() 13`] = `"DAY_ONLY(LastLogin)"`;

exports[`Fn #toString() 14`] = `"FISCAL_MONTH(LastLogin)"`;

exports[`Fn #toString() 15`] = `"FISCAL_QUARTER(LastLogin)"`;

exports[`Fn #toString() 16`] = `"FISCAL_YEAR(LastLogin)"`;

exports[`Fn #toString() 17`] = `"HOUR_IN_DAY(LastLogin)"`;

exports[`Fn #toString() 18`] = `"MAX(max)"`;

exports[`Fn #toString() 19`] = `"MIN(min)"`;

exports[`Fn #toString() 20`] = `"SUM(sum)"`;

exports[`Fn #toString() 21`] = `"WEEK_IN_MONTH(LastLogin)"`;

exports[`Fn #toString() 22`] = `"WEEK_IN_YEAR(LastLogin)"`;
